eval/src/vespa/eval/instruction/generic_peek.cpp

```diff
@@ -0,0 +1,280 @@
+
+using State = InterpretedFunction::State;
+using Instruction = InterpretedFunction::Instruction;
+
+namespace {
+
+static constexpr size_t npos = -1;
+
+enum class Source { COPY_INPUT, FROM_CHILD, VERBATIM };
+
+struct DenseSizes {
+    std::vector<size_t> loop_cnt;
+    std::vector<size_t> stride;
+
+    DenseSizes(const ValueType &input_type)
+        : loop_cnt(), stride()
+    {
+        for (const auto &dim : input_type.dimensions()) {
+            if (dim.is_indexed()) {
+                loop_cnt.push_back(dim.size);
+            }
+        }
+        stride.resize(loop_cnt.size());
+        size_t out_size = 1;
+        for (size_t i = loop_cnt.size(); i-- > 0; ) {
+            stride[i] = out_size;
+            out_size *= loop_cnt[i];
+        }
+        assert(out_size == input_type.dense_subspace_size());
+    }
+};
+
+/** Compute input offsets for all output cells */
+struct DensePlan {
+    std::vector<size_t> loop_cnt;
+    std::vector<size_t> in_stride;
+    size_t verbatim_offset = 0;
+    std::vector<size_t> child_idx;
+    std::vector<size_t> child_stride;
+    std::vector<size_t> child_limit;
+
+    DensePlan() = default;
+
+    void init_from(const DenseSizes &sizes,
+                   const std::vector<std::pair<Source,size_t>> &where)
+    {
+        for (size_t i = 0; i < where.size(); ++i) {
+            auto [from, idx] = where[i];
+            switch (from) {
+            case Source::COPY_INPUT:
+                loop_cnt.push_back(sizes.loop_cnt[i]);
+                in_stride.push_back(sizes.stride[i]);
+                break;
+            case Source::VERBATIM:
+                assert(idx < sizes.loop_cnt[i]);
+                verbatim_offset += idx * sizes.stride[i];
+                break;
+            case Source::FROM_CHILD:
+                child_stride.push_back(sizes.stride[i]);
+                child_idx.push_back(idx);
+                child_limit.push_back(sizes.loop_cnt[i]);
+                break;
+            }
+        }
+    }
+
+    /** Get initial offset (from verbatim labels and child values) */
+    template <typename Getter>
+    size_t get_offset(const Getter &get_child_value) const
+    {
+        size_t offset = verbatim_offset;
+        for (size_t i = 0; i < child_idx.size(); ++i) {
+            size_t from_child = get_child_value(child_idx[i]);
+            if (from_child < child_limit[i]) {
+                offset += from_child * child_stride[i];
+            } else {
+                return npos;
+            }
+        }
+        return offset;
+    }
+
+    template<typename F> void execute(size_t offset, const F &f) const {
+        run_nested_loop<F>(offset, loop_cnt, in_stride, f);
+    }
+};
+
+using SparsePlan = std::vector<std::pair<Source, TensorSpec::Label>>;
+
+struct PeekParam {
+    const ValueType res_type;
+    size_t out_mapped_dims;
+    size_t out_dense_size;
+    const ValueType input_type;
+    size_t in_dense_size;
+    DenseSizes dense_sizes;
```

Comment:
maybe dense_sizes and dense_where could be temporary variables that are not part of the actual Param object since they seem to only be used to initialize the dense plan.